src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,444 @@
+            for (auto [idx, val] : lut) {
+                writeBinaryPOD(writer, idx);
+                writeBinaryPOD(writer, val);
+            }
+        }
+        if (use_wand_) {
+            for (size_t i = 0; i < n_cols_internal(); ++i) {
+                writeBinaryPOD(writer, max_in_dim_[i]);
+            }
+        }
+        return Status::success;
+    }
+
+    Status
+    Load(MemoryIOReader& reader) {
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        readBinaryPOD(reader, use_wand_);
+        readBinaryPOD(reader, drop_ratio_build_);
+        size_t rows;
+        readBinaryPOD(reader, rows);
+        raw_data_.resize(rows);
+        size_t dim;
+        readBinaryPOD(reader, dim);
+        for (size_t i = 0; i < rows; ++i) {
+            auto& row = raw_data_[i].data;
+            size_t len;
+            readBinaryPOD(reader, len);
+            row.resize(len);
+            for (size_t j = 0; j < len; ++j) {
+                readBinaryPOD(reader, row[j].first);
+                readBinaryPOD(reader, row[j].second);
+            }
+        }
+
+        inverted_lut_.resize(dim);
+        for (size_t i = 0; i < dim; ++i) {
+            size_t lut_size;
+            readBinaryPOD(reader, lut_size);
+            inverted_lut_[i].resize(lut_size);
+            for (size_t j = 0; j < lut_size; ++j) {
+                readBinaryPOD(reader, inverted_lut_[i][j].id);
+                readBinaryPOD(reader, inverted_lut_[i][j].distance);
+            }
+        }
+        if (use_wand_) {
+            max_in_dim_.resize(dim);
+            for (size_t i = 0; i < dim; ++i) {
+                readBinaryPOD(reader, max_in_dim_[i]);
+            }
+        }
+        return Status::success;
+    }
+
+    Status
+    Add(const SparseRow<T>* data, size_t rows, int64_t dim) {
+        std::unique_lock<std::shared_mutex> lock(mu_);
+        auto current_rows = n_rows_internal();
+        if (current_rows != 0 && drop_ratio_build_ != 0) {
+            LOG_KNOWHERE_ERROR_ << "Inserting to non empty sparse index with non zero drop_ratio_build is not allowed.";
+            return Status::invalid_args;
+        }
+        auto val_threshold = 0.0f;
+        if (drop_ratio_build_ != 0.0f) {
+            auto global_max = -std::numeric_limits<T>::max();
+            for (size_t i = 0; i < rows; ++i) {
+                for (auto [idx, val] : data[i].data) {
+                    global_max = std::max(global_max, val);
+                }
+            }
+            val_threshold = global_max * drop_ratio_build_;
+        }
+
+        if (inverted_lut_.size() < dim) {
+            inverted_lut_.resize(dim);
+            if (use_wand_) {
+                max_in_dim_.resize(dim);
+            }
+        }
+
+        raw_data_.resize(current_rows + rows);
+        for (size_t i = 0; i < rows; ++i) {
+            raw_data_[current_rows + i] = data[i];
+            for (auto [idx, val] : data[i].data) {
+                if (val < val_threshold) {
```

Comment:
if `drop_ratio_build_ == 0`, then `val_threshold` is `0`, so negative `val` values will be skipped. Is it by design? This implies that sparse values are expected to be positive only

Reply:
Thanks Alex! I have changed it so that now we ignore only values that are close enough to zero.

Also the drop strategy has been modified from "drop any value that is below ratio * global max" to "drop the smallest ratio% abs values".